sys/hashes/sha512.c

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (C) 2023 TU Dresden
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     sys_hashes
+ * @{
+ *
+ * @file
+ * @brief       SHA-512 hash function implementation
+ *
+ * @author      Mikolai Gütschow <mikolai.guetschow@tu-dresden.de>
+ *
+ * @}
+ */
+
+#include <string.h>
+#include <assert.h>
+
+#include "hashes/sha512.h"
+#include "hashes/sha512_common.h"
+
+/* SHA-512 initialization.  Begins a SHA-512 operation. */
+void sha512_init(sha512_context_t *ctx)
+{
+    /* Zero bits processed so far */
+    ctx->count[0] = ctx->count[1] = 0;
+
+    /* Magic initialization constants */
+    ctx->state[0] = 0x6a09e667f3bcc908;
+    ctx->state[1] = 0xbb67ae8584caa73b;
+    ctx->state[2] = 0x3c6ef372fe94f82b;
+    ctx->state[3] = 0xa54ff53a5f1d36f1;
+    ctx->state[4] = 0x510e527fade682d1;
+    ctx->state[5] = 0x9b05688c2b3e6c1f;
+    ctx->state[6] = 0x1f83d9abfb41bd6b;
+    ctx->state[7] = 0x5be0cd19137e2179;
+}
+
+void *sha512(const void *data, size_t len, void *digest)
+{
+    sha512_context_t c;
+    static unsigned char m[SHA512_DIGEST_LENGTH];
```

Comment:
This looks like a hard-to-debug footgun waiting to happen. I think it is better to either mutex the structure if you want to keep it static, or remove the `static` and make it a local variable.

Reply:
Second option doesn't work since a pointer to it is returned.

A mutex will help for _immediate_ data races, but still cannot save the user of concurrent access issues to the `sha512()` since the content of the buffer might be updated by the time the calling thread checks the buffer.
Maybe it is just better to remove the static buffer altogether? Same applies to `sha256()` then, btw.

Reply:
After reading the code for a second time I see how exactly this is used and why my proposed solutions are not correct.

I would remove the variable and the option to pass a NULL as `digest` altogether and change the return type to a `void`. This way the caller always has to supply storage for the digest. I didn't check the PSA code part, so I don't know if that part of the code uses the feature. I'd like to have @Teufelchen1's opinion on this too.

Reply:
See also #20116 where the same pattern is removed from the sha256 code base

Reply:
I think its a good idea to remove that behavior.
Unlike* to the change in #20116, the unit tests in this PR would catch that. So @mguetschow will need to touch them once again.

*(This is a repeating failure within RIOT, we *must* increase the quality of our unit tests!)

Reply:
Done in 29d32b1. Tell me when to squash.

Reply:
> Done in [29d32b1](https://github.com/RIOT-OS/RIOT/commit/29d32b14544bd414d64fed2036c8e2b743ca352e). Tell me when to squash.

Please squash!

Reply:
Done :)